Allow adding permisisons and identites when creatign a group. Simplify group editing

edlerd opened the pull request:
## Done

- Allow adding permissions and identities when creating a group.
- Simplify group editing

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - create a new permission group
    - edit permission group: 1. add identities, 2. remove identities 3. add permissions 4. remove permissions

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://lxd-ui-887.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/14783)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

edlerd commented:
Addressed 2, 3, 4 and 5. I think the others need a bit more discussion.

mas-who commented on src/pages/permissions/panels/EditGroupPermissionsForm.tsx:
```diff
@@ -0,0 +1,275 @@
+  isEditing?: boolean;
+}
+
+const EditGroupPermissionsForm: FC<Props> = ({
```
love the new approach with `isAdded` and `isRemoved` :)

mas-who commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -11,26 +17,110 @@ import { renameGroup, updateGroup } from "api/auth-groups";
+    const identityPayload: LxdIdentity[] = [];
+
+    const existingIdentities = await fetchIdentities();
+    existingIdentities.map((identity) => {
```
again I think the logic here can be alot cleaner if we include api data in `identities` state.

mas-who commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -11,26 +17,110 @@ import { renameGroup, updateGroup } from "api/auth-groups";
+      .filter((id) => id.isRemoved)
+      .map((id) => id.id);
+
+    const identityPayload: LxdIdentity[] = [];
```
I think we should skip the rest of the logic here if `addedIdentities` and `removedIdentities` are both empty.

mas-who commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -39,78 +129,141 @@ const EditGroupPanel: FC<Props> = ({ group }) => {
+      const isPermissionRemoved = permissions.find((p) => p.isRemoved);
+      if (isSelfRemoved || (isSelfInGroup && isPermissionRemoved)) {
+        setConfirming(true);
+      } else {
```
nice one!

mas-who commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -39,78 +129,141 @@ const EditGroupPanel: FC<Props> = ({ group }) => {
+          {subForm === "identity" && (
+            <SidePanel.HeaderTitle>
+              <Icon name="chevron-left" />{" "}
+              <Button
```
same comment about the header title component as in `CreateGroupPanel`

mas-who commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -39,78 +129,141 @@ const EditGroupPanel: FC<Props> = ({ group }) => {
+              permissions={permissions}
+              setPermissions={(value) => setPermissions(value)}
+              group={group}
+            />
```
```suggestion
              setPermissions={setPermissions}
```

mas-who commented on src/pages/permissions/actions/GroupActions.tsx:
```diff
@@ -13,93 +12,35 @@ interface Props {
-              </>
-            ),
-          },
+      <List
```
this is very nice to see, so much code removed :)

edlerd commented on src/pages/permissions/panels/CreateGroupPanel.tsx:
```diff
@@ -60,37 +104,102 @@ const CreateGroupPanel: FC = () => {
+            <SidePanel.HeaderTitle>Create group</SidePanel.HeaderTitle>
+          )}
+          {subForm === "identity" && (
+            <SidePanel.HeaderTitle>
```
Yes good idea! We can make this much less repetitive, by extracting this bit into a dedicated GroupHeaderTitle component.

edlerd commented on src/pages/permissions/forms/GroupForm.tsx:
```diff
@@ -10,9 +11,23 @@ export interface GroupFormValues {
 
 interface Props {
   formik: FormikProps<GroupFormValues>;
+  setSubForm: (value: string) => void;
```
I'll a dedicated GroupSubForm type for this.

edlerd commented on src/pages/permissions/panels/CreateGroupPanel.tsx:
```diff
@@ -60,37 +104,102 @@ const CreateGroupPanel: FC = () => {
+                setSubForm={setSubForm}
+                identityCount={identities.length}
+                identityModifyCount={identities.length}
+                permissionCount={permissions.filter((p) => !p.isRemoved).length}
```
With the new logic of removing those on creation instead of setting them as removed, this simplifies to not need a filter. similar to the identitites.length above.

edlerd commented on src/pages/permissions/panels/CreateGroupPanel.tsx:
```diff
@@ -12,13 +17,22 @@ import { queryKeys } from "util/queryKeys";
+import { FormPermission } from "pages/permissions/panels/PermissionSelector";
+import { fetchIdentities, updateIdentities } from "api/auth-identities";
+import EditGroupPermissionsForm from "pages/permissions/panels/EditGroupPermissionsForm";
+import { FormIdentity } from "pages/permissions/panels/EditGroupPanel";
```
right, good point :+1:

edlerd commented on src/pages/permissions/panels/CreateGroupPanel.tsx:
```diff
@@ -12,13 +17,22 @@ import { queryKeys } from "util/queryKeys";
   const queryClient = useQueryClient();
   const controllerState = useState<AbortController | null>(null);
+  const [subForm, setSubForm] = useState<string | null>(null);
+  const [identities, setIdentities] = useState<FormIdentity[]>([]);
```
Yes, that simplifies things down the road, though it complicates the loading of the edit form a bit. Applied this change.

edlerd commented on src/pages/permissions/forms/PanelFormLink.tsx:
```diff
@@ -0,0 +1,48 @@
+      onClick={() => onClick(entity)}
+      type="button"
+    >
+      <span className="panel-form-link__column">
```
I tried this, but it seems to make this more messy than with custom styling. I think row and col are useful when aligning page content, it is not meant to align elements of a button.

mas-who replied:
cool, let's leave it as is then :+1:

edlerd commented:
Addressed all the comments, and the design points 1. and 7.

Only remaining is 6. which we probably best discuss in person. In short the problem with undo is when switching between forms, the a generic undo stack for all 3 forms would be confusing and having 3 different undo stacks seems confusing as well.

mas-who commented:
> Addressed all the comments, and the design points 1. and 7.
> 
> Only remaining is 6. which we probably best discuss in person. In short, the problem with undo is when switching between forms (a new interaction pattern for those panels), a generic undo stack for all 3 forms would be confusing and having 3 different undo stacks seems confusing as well.

Thanks! I was thinking about 6. as I reviewed earlier, like you said I think it's not worth the complexity to include undo/redo as a feature here, let's leave it out :+1:

mas-who commented:
QA looks mostly good! Just one more thing, seems the "Cancel" and "Create group" buttons are still hidden when adding identities or permissions in the create group panel.

mas-who commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -11,26 +16,128 @@ import { renameGroup, updateGroup } from "api/auth-groups";
+  const [subForm, setSubForm] = useState<GroupSubForm>(null);
+  const [confirming, setConfirming] = useState(false);
+  const [identities, setIdentities] = useState<FormIdentity[]>(
+    getIdentityIdsForGroup(group).map((id) => ({ id: id })) as FormIdentity[],
```
this is a little unfortunate, but I think it's worth the simplification everywhere else...

mas-who commented:
thanks for the changes, QA and code both LGTM :+1:

edlerd commented:
@piperdeck can you have a look at this one? It tries to improve the experience on creation and editing of permission groups.

mas-who commented:
The delete button for groups is cut off at the edge on smaller screen sizes.

![Screenshot from 2024-09-12 14-49-40](https://github.com/user-attachments/assets/4a11e88e-0dec-403d-b0ed-358be86b33bb)

edlerd commented:
> The delete button for groups is cut off at the edge on smaller screen sizes.

Good catch! Added a fix for it just now.

mas-who reviewed (commented):
Thanks for the improvement! Very minor code comments only. QA looks good!

edlerd commented on src/pages/permissions/panels/EditGroupPanel.tsx:
```diff
@@ -11,26 +16,159 @@ import { renameGroup, updateGroup } from "api/auth-groups";
+    queryFn: () => fetchImageList(),
+  });
+  useEffect(() => {
+    if (!identityLoading && !imageLoading) {
```
good idea, added the error result to the identity and image loading and checking against it in the use effects.

mas-who commented:
LGTM, again :slightly_smiling_face:
